bitswap/network/http_multiaddr.go

```diff
@@ -0,0 +1,128 @@
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+)
+
+// ParsedURL contains the result of parsing an "http" transport multiaddress.
+// SNI is set when the multiaddress specifies an SNI value.
+type ParsedURL struct {
+	URL *url.URL
+	SNI string
+}
+
+// ExtractHTTPAddress extracts the HTTP schema+host+port from a multiaddress
+// and returns a *url.URL and an SNI string if present.
+func ExtractHTTPAddress(ma multiaddr.Multiaddr) (ParsedURL, error) {
+	components := ma.Protocols()
+	var host, port, schema, sni string
+	var tls bool
+
+	for _, comp := range components {
+		switch comp.Name {
+		case "dns", "dns4", "dns6", "ip4", "ip6":
+			hostVal, err := ma.ValueForProtocol(comp.Code)
+			if err != nil {
+				return ParsedURL{}, fmt.Errorf("failed to extract host: %w", err)
+			}
+			host = hostVal
+		case "tcp", "udp":
+			portVal, err := ma.ValueForProtocol(comp.Code)
+			if err != nil {
+				return ParsedURL{}, fmt.Errorf("failed to extract port: %w", err)
+			}
+			port = portVal
+		case "tls":
+			tls = true
+		case "http":
+			schema = "http"
+			if tls {
+				schema = "https"
+			}
+		case "https":
+			schema = "https"
+		case "sni":
```

Comment:
`sni` probably should work the same way as `tls` and force `https`. unlikely, but just in case someone misconfigured things and announced `/sni/`without `/tls/`
```suggestion
		case "sni":
		  schema = "https"
```

Reply:
is this spec'ed somewhere?

Reply:
3ea55a6ddc163561c15b2e57e5565f1eb6e17e06